### ISSUE TITLE
Design Tokens - Requirements gathering

### ISSUE THREAD

**Issue by tomheller** [dynatrace-oss/barista#1057]
We are planning to include Design Tokens into our library for everyone to use. If you are not familiar with design tokens, you can use these resources to familiarize yourself with the subject, but the gist of it is to have a central definition for every spacing / color / border / radius / opacity / font / ... used within the design language to make the design more consistent.

* https://spectrum.adobe.com/page/design-tokens/
* https://css-tricks.com/what-are-design-tokens/
* https://www.lightningdesignsystem.com/design-tokens/

We already have started toying with design tokens and you can find our [concept specification within the repo](https://github.com/dynatrace-oss/barista/blob/master/concepts/design-tokens/README.md).

So far, we are have a couple of design tokens defined with dummy values like spacing, colors and some for the web-component based button. These design tokens currently generate the following output variable(formats):

**sass variables**
Variable definitions that can be used within your sass/scss environment, for example: `$palette-color-key-100`. 

**css custom properties**
Variable definitions that can be used within your css environment, for example: `var(--palette-color-key-100)`. 

**Typescript constants**
Currently for completions sake. Sometimes one does need a specific color or spacing definition in your scripts and it is good to know that we can keep them in sync via the design tokens definition, for example: `export const PALETTE_COLOR_KEY_100='#ff00ff';`

## Help wanted
As we are ramping up speed with the design tokens, we now want your input on what you would expect from these. What outputs would help you to use the tokens within your (angular) application / prototyping setup / ect...

Please let us know and we will definitely consider your needs.

**Comment by ffriedl89** [comment]
We were debating whether utility classes ala tailwind make sense for us. 
For spacings this would mean that we would need to generate a lot of css classes. Without proper tooling (purgeCSS for unused classes) the amount of css code shipped can get quite out of hand. 
e.g. for spacing-xl we would need to create 
```
// padding
.dt-p-xl
// padding top
.dt-p-t-xl
.dt-p-r-xl
.dt-p-b-xl
.dt-p-l-xl

// margin
.dt-m-xl
.dt-m-t-xl
.dt-m-r-xl
.dt-m-b-xl
.dt-m-l-xl
```

So roughly 10 css classes for one spacing token. If unused and not removed from the bundle this gets really big. 
Another approach would be to generate mixins that can be used by the consumer. These mixins don't generate code unless used which would remove the bloating problem.

**Comment by ffriedl89** [comment]
After rethinking the scaling and bundle size, I think generating those helper classes for certain tokens should be good. Spacings for example should be rather small compared to colors for example. Colors have the issue that you would need classes for, background-colors, border-colors, font-colors, hover & active styles and with the amount of colors that we have this would result in an insane amount of generated classes. 

@ddprrt What do you think about having colors only as css custom properties available, but spacings and maybe other tokens as helper classes?

**Comment by ddprrt** [comment]
Ad Utility classes approach: I think we have to see Tailwind and utility class-based frameworks from a different angle. Not as an outlet providing design tokens, but as a design system as a whole. Tailwind's classes are pretty much defined, there is no room and no need for any other classes, everything else is just composition. This makes the CSS output constant. No matter which application you create. Related: https://www.youtube.com/watch?v=R50q4NES6Iw

I would love to see what happens if we use our design system (or tokens) and apply them to Tailwind. Let us discuss if we can do a research on that.

Ad colors. Agreed. With our rich color palettes and the option to get even more color palettes, this might be beyond the point. It might make more sense to focus on utility classes for themes: https://barista.dynatrace.com/patterns/theming and set palettes (using the cascade)

**Comment by tomheller** [comment]
@ddprrt Toyed around with your idea a little bit, and I am positively impressed on what we were able to do with the configuration.
The css file has been generated based on the config enclosed. 
Be aware that this only partially resembles our design tokens as colors and fonts are still in flux, but maybe you can already play around with it. 

https://codesandbox.io/s/proud-shadow-6j6og?file=/index.html

**Comment by ddprrt** [comment]
Cool, this is exactly what I've imagined! Guess we can call this our first deliverable 😎

**Comment by tomheller** [comment]
Automatic generation and embedding within the published package was done in #1199 and #1220
